binary-oracle-pair/program/tests/tests.rs

```diff
@@ -0,0 +1,675 @@
+
+    let user_fail_tokens = get_token_balance(&mut banks_client, &user_fail_account.pubkey()).await;
+    assert_eq!(user_fail_tokens, deposit_amount);
+}
+
+#[tokio::test]
+async fn test_withdraw() {
+    let (mut banks_client, payer, recent_blockhash) = program_test().start().await;
+
+    let deposit_amount = 100;
+
+    let pool = TestPool::create();
+
+    pool.init_pool(&mut banks_client, &payer, &recent_blockhash)
+        .await;
+
+    let user_account = Keypair::new();
+    let user_account_owner = Keypair::new();
+    let user_pass_account = Keypair::new();
+    let user_fail_account = Keypair::new();
+
+    pool.prepare_accounts_for_deposit(
+        &mut banks_client,
+        &payer,
+        &recent_blockhash,
+        deposit_amount,
+        deposit_amount,
+        &user_account,
+        &user_account_owner,
+        &user_pass_account,
+        &user_fail_account,
+    )
+    .await;
+
+    // Make deposit
+    pool.make_deposit(
+        &mut banks_client,
+        &payer,
+        &recent_blockhash,
+        &user_account,
+        &user_pass_account,
+        &user_fail_account,
+        deposit_amount,
+    )
+    .await;
+
+    // Set allowances to burn PASS and FAIL tokens
+    approve_delegate(
+        &mut banks_client,
+        &payer,
+        &recent_blockhash,
+        &user_pass_account.pubkey(),
+        &pool.authority,
+        &user_account_owner,
+        deposit_amount,
+    )
+    .await
+    .unwrap();
+    approve_delegate(
+        &mut banks_client,
+        &payer,
+        &recent_blockhash,
+        &user_fail_account.pubkey(),
+        &pool.authority,
+        &user_account_owner,
+        deposit_amount,
+    )
+    .await
+    .unwrap();
+
+    let user_balance_before = get_token_balance(&mut banks_client, &user_account.pubkey()).await;
+    assert_eq!(user_balance_before, 0);
+
+    // Check balance of pool deposit account
+    let pool_deposit_account_balance =
+        get_token_balance(&mut banks_client, &pool.pool_deposit_account.pubkey()).await;
+    assert_eq!(pool_deposit_account_balance, deposit_amount);
+
+    // Check if user has PASS and FAIL tokens
+    let user_pass_tokens = get_token_balance(&mut banks_client, &user_pass_account.pubkey()).await;
+    assert_eq!(user_pass_tokens, deposit_amount);
+
+    let user_fail_tokens = get_token_balance(&mut banks_client, &user_fail_account.pubkey()).await;
+    assert_eq!(user_fail_tokens, deposit_amount);
+
+    let mut transaction = Transaction::new_with_payer(
+        &[instruction::withdraw(
+            &id(),
+            &pool.pool_account.pubkey(),
+            &pool.authority,
+            &pool.pool_deposit_account.pubkey(),
+            &user_pass_account.pubkey(),
+            &user_fail_account.pubkey(),
+            &pool.token_pass_mint.pubkey(),
+            &pool.token_fail_mint.pubkey(),
+            &user_account.pubkey(),
+            &spl_token::id(),
+            deposit_amount,
```

Comment:
This test can be modified to withdraw some amount less than `deposit_amount` to test the overdraw issue pointed out before